texastribune/scuole
# Disallow unknown hosts in nginx

## What does this PR do?
- Motivation: DV team gets a lot of emails "Invalid HTTP_HOST header". This error is coming from Django. We need to disallow these requests at the nginx level. See more in the ticket [here](https://airtable.com/appyo1zuQd8f4hBVx/tbloNZu8GkM52NKFR/viwXlAz9IbHE4sbIX/rec6Q55t0KgabFb7P?blocks=hide).

- This PR adds nginx configuration to disallow any request whose host does not include "anysubdomain.texastribune.org". 

## Any relevant ticket
- [ticket](https://airtable.com/appyo1zuQd8f4hBVx/tbloNZu8GkM52NKFR/viwXlAz9IbHE4sbIX/rec6Q55t0KgabFb7P?blocks=hide)

## Difficulty in testing before production
- It is difficult to test this implementation even in staging server because cloudflare blocks it. This error was showing up precisely because the hackers/bots/requesters were able to bypass host requirements from cloudflare.
   -  ![image](https://github.com/texastribune/scuole/assets/119520346/8f928815-1403-4041-b2c6-a4987346833a)

   - We can still test it once we deploy within staging server (let me know if this would not work)
      - regular request from browser
      - ssh into the staging instance and send a curl request
 
## How to test it locally?

- Option 1 to test: I made a separate directory to test this. See this [commit](https://github.com/texastribune/scuole/pull/936/commits/9267e19808b6c051fc9626a3f14cba86f551dcb8). You can checkout to this commit and run below (I kept these commands as comments in dockerfile)
```sh
docker stop testserver
docker build -t nginx-m-servers .
docker run --rm --name testserver -d -p 80:80 nginx-m-servers
curl localhost:80 -H "Host: example.com"
curl localhost:80 -H "Host: abc.example.com"
curl localhost:80 -H "Host: abc.nexample.com"
```

- Option 2 to test
   - Make a directory and cd into it.
   - Add a new file named default and [copy these codes](https://github.com/texastribune/scuole/blob/9267e19808b6c051fc9626a3f14cba86f551dcb8/test-case-nginx/default). Line 8 and 9 are our interest.
   - Add another new file, Dockerfile, and copy the codes below
```sh
#  Pull the minimal Ubuntu image
FROM nginx

# Copy the Nginx config
COPY default /etc/nginx/conf.d/default.conf
``` 
   - Run below
```sh
docker stop testserver
docker build -t nginx-m-servers .
docker run --rm --name testserver -d -p 80:80 nginx-m-servers
curl localhost:80 -H "Host: example.com" 
curl localhost:80 -H "Host: abc.example.com"
curl localhost:80 -H "Host: abc.nexample.com"
``` 

   - Expected results
      - ![image](https://github.com/texastribune/scuole/assets/119520346/1c9d7621-df3c-4b62-aa58-c283719ef6cb)


## How to deploy it?
- Merge this code
- Deploy in schools test and see if it working fine there
- Deploy in schools prod 1 and see if it working fine there
- Deploy in schools prod 2 and see if it working fine there
- Be ready to reverse this change
   - 1. edit the nginx default and run nginx reload
   - 2. Reverse the code, merge, deploy

## How do we know it succeeded
- Work with DV team to check if these error email has stopped

## Any backup plans?
- Be ready to reverse this change.

## Any Tech-debt
- NA
